### ISSUE TITLE
KOReader freezes after predetermined time (Appimage)

### ISSUE THREAD

**Issue by offset-torque** [koreader/koreader#9262]
* KOReader version: koreader-appimage-x86_64-linux-gnu-v2022.06
* Device: Linux PC - KDE - X11

#### Issue
KOReader freezes after:
- 5 minutes if menu is open
```
06/27/22-16:56:39 INFO  Loading plugins from directory
06/27/22-17:11:40 INFO  Inhibiting user input
```

- 15 minutes if menu is not open
```
06/27/22-18:10:11 INFO  Loading plugins from directory
06/27/22-18:25:12 INFO  Inhibiting user input
```

Exact times and the inhibition message makes me think that this might be intentional.
But this puts KOReader into an unresponsive state.
No mouse or keyboard input is registered.
Even the close button doesn't work.
It has to be killed from the task manager or terminal.

#### Steps to reproduce
- Open KOReader
- Don't click on anything for the durations mentioned above

**Comment by Frenzie** [comment]
You don't see a little suspended message? (Not that you should in a regular desktop app, but it's probably something from the emulator.)

**Comment by NiLuJe** [comment]
Verbose debug, please ;).

Because, yeah, that looks like AutoSuspend attempting a suspend, which should either not happen at all or do something visible in the form of the suspend screensaver as @Frenzie mentioned ;).

**Comment by NiLuJe** [comment]
Even if you *did* see the screensaver, we're probably missing an SDL-specific keycode whitelisted to get out of the simulated suspend...

Thoughts?

**Comment by NiLuJe** [comment]
Hmm, the SDL event handler is buried in the SDL implementation, it might just be easier to simply *not* void the SDL handler during suspend...

**Comment by Frenzie** [comment]
It used to be that performing an action canceled out suspend on the emulator. You or @zwim changed that recently for reasons that are unclear to me. ;-)

**Comment by Frenzie** [comment]
But I think F2 should still get out of it, probably?

**Comment by NiLuJe** [comment]
Okay, yeah, the SDL event handler appears to be for, well, SDL-specific bits.

I'm *hoping* the rest goes through standard EV_KEY stuff, so, yeah, F2 might do the trick?

Regardless of that, nerfing that handler was probably not a great idea ;p.

(Waiting on verbose debug logs to confirm).

**Comment by offset-torque** [comment]
> You don't see a little suspended message?

Nope, it freezes as whatever on the screen at the moment of inhibition message, even the menu is visible

> But I think F2 should still get out of it, probably?

Nope, nothing works including F2

---

Apparently it is not frozen internally.

After getting this message:
`INFO  Inhibiting user input`

**KOReader stays unresponsive while these keycodes are scrolling in the terminal:**

**Mouse input (trying to open menu)**
```
DEBUG input event => type: 3 (EV_ABS), code: 57 (ABS_MT_TRACKING_ID), value: 0
DEBUG input event => type: 3 (EV_ABS), code: 53 (ABS_MT_POSITION_X), value: 169
DEBUG input event => type: 3 (EV_ABS), code: 54 (ABS_MT_POSITION_Y), value: 395
DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0
DEBUG input event => type: 3 (EV_ABS), code: 57 (ABS_MT_TRACKING_ID), value: -1
DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0
DEBUG input event => type: 83 (EV_SDL), code: 1027, value: cdata<struct SDL_MouseWheelEvent &>: 0x7eff386789d8
```

**F2 (which is exactly same whether it is frozen or not)**
```
DEBUG key event => code: 1073741883 (Power), value: 1
DEBUG AutoSuspend: onInputEvent
DEBUG AutoTurn: onInputEvent
DEBUG key event => code: 1073741883 (Power), value: 0
DEBUG AutoSuspend: onInputEvent
DEBUG AutoTurn: onInputEvent
```

**Close button**
```
DEBUG key event => code: 1073742050 (Alt), value: 1
DEBUG key event => code: 1073741885 (F4), value: 1
```

**Comment by NiLuJe** [comment]
> > You don't see a little suspended message?
> 
> Nope, it freezes as whatever on the screen at the moment of inhibition message, even the menu is visible

That's... weird. Another issue, possibly, mildly related, but weird.

**Comment by yparitcher** [comment]
I bumped into this today while trying to test suspend in the emulator.

I may have abused koreader last night to write a custom dynamic kindle screensaver ;) (Koreader is a better gui toolkit than FBInk)

related:
Does anyone mind if i implement `WakeupManager` for kindle? On `supportsScreensaver` and non legacy devices?

This would be by listening for `powerd`'s `wakeup` & `readyToSuspend` events, and setting the rtc when in the `readyToSuspend` state via lipc.

**Comment by NiLuJe** [comment]
Go for it ;).

**Comment by Frenzie** [comment]
@yparitcher 
Not sure why we'd mind. :-)

@NiLuJe
> Okay, yeah, the SDL event handler appears to be for, well, SDL-specific bits.

I'm not really sure what you mean by the SDL event handler. Are you talking about the part where I decided that handling some SDL events more intelligently in front made a lot more architectural sense than trying to do a bunch of pretend down in base?

**Comment by NiLuJe** [comment]
> > Okay, yeah, the SDL event handler appears to be for, well, SDL-specific bits.
> 
> I'm not really sure what you mean by the SDL event handler. Are you talking about the part where I decided that handling some SDL events more intelligently in front made a lot more architectural sense than trying to do a bunch of pretend down in base?

Nah, I just misremembered how the whole SDL thing worked, as far as input event goes. There's much less SDL-specific bits for "pure" input than I originally thought (thanks to said base shenanigans), so the stuff in front is mostly not-really-input anymore ;).

Still, the point stands, might not be great to inhibit those, I'll check it out later ;).

**Comment by Frenzie** [comment]
I wrote that, yes. But the way "suspend" is supposed to work on emulator in my view (not on regular desktop but presumably no one ever changed it) is that it just shows a little popup or whatever UI saying hello, we're in suspend status, and basically any input cancels it. That suffices for the UI and you can't develop the suspend without real hardware anyway.

> There's much less SDL-specific bits for "pure" input than I originally thought (thanks to said base shenanigans), so the stuff in front is mostly not-really-input anymore ;).

It depends a bit on the specifics what makes sense where but I may well intend to lift more handling stuff to front 'cause in base you're very limited to a certain kind of emulation instead of directly tapping into things. Which was perfectly acceptable for the original emulator concept of course. ;-)

**Comment by NiLuJe** [comment]
Okay, it works as it ought to with #9263 applied over here.

I'm *definitely* seeing the screensaver, whether I trip it via Exit > Sleep, or via AutoSuspend.

![emu_sleep](https://user-images.githubusercontent.com/111974/176244632-e68760f9-fb9a-4e3b-8fae-a708e8734a3f.png)
